icevision/models/mmdet/common/bbox/single_stage/model.py

```diff
@@ -5,26 +5,49 @@
 from mmdet.models import build_detector
 from mmcv.runner import load_checkpoint
 from icevision.models.mmdet.utils import *
+from icevision.utils.download_utils import *
 
 
 def model(
-    cfg: Union[str, Path, Config],
+    backbone,
     num_classes: int,
-    weights_path: Optional[Union[str, Path]] = None,
+    pretrained: bool = True,
+    checkpoints_path: Optional[Union[str, Path]] = None,
+    force_download=False,
 ) -> nn.Module:
 
-    # if `cfg` argument is a path (str, Path) create an Config object from the file
-    # otherwise cfg should be already an Config object
-    if isinstance(cfg, (str, Path)):
-        cfg = Config.fromfile(str(cfg))
+    save_dir = None
+
+    model_name = backbone["model_name"]
+    cfg_filepath = backbone["cfg_filepath"]
+    weights_url = backbone["weights_url"]
+
+    # download weights
+    if pretrained and weights_url:
+        if checkpoints_path:
+            save_dir = checkpoints_path.model_name
+        else:
+            save_dir = Path("checkpoints") / model_name
+        save_dir.mkdir(exist_ok=True, parents=True)
+
+        fname = Path(weights_url).name
+        weights_path = save_dir / fname
+
+        if not weights_path.exists() or force_download:
+            download_url(weights_url, weights_path)
+
+    if cfg_filepath:
+        cfg = Config.fromfile(cfg_filepath)
+    else:
```

Comment:
I'm not sure we need `if cfg_filepath`, if this is an invalid path the function `Config.fromfile` will already thrown an error I guess

Reply:
We are checking if the path is empty then we raise an error with a meaningful error message. I replace the message by the following one:

```
raise RuntimeError(
            "cfg_filepath is empty. Model must have a valid configuration file"
        )
```

Reply:
What error message do we get if you pass an invalid path to `Config.fromfile`?

Reply:
```
def check_file_exist(filename, msg_tmpl='file "{}" does not exist'):
    if not osp.isfile(filename):
        raise FileNotFoundError(msg_tmpl.format(filename))

```
Basically, if the filename is empty the mmcv message will be:
**file  does not exist**
not very helpful message there. As opposed to the one I'm adding:
**cfg_filepath is empty. Model must have a valid configuration file**

Reply:
I'm not sure the modified error message is easier to understand. It's not clear what "cfg_filepath is empty" means, do we have a file but it's empty? 

Probably I'm not understading here, but I'm guessing the problem really is when you pass a Path that does not exist, in this case the original error message makes more sense.

I also thought that by "cfg_filepath is empty" you could also be referring that it was not set on `MMDetBackboneConfig`. But you can't instantiate `MMDetBackboneConfig` without specifying a `cfg_filepath`

Reply:
`Raise Error` removed: handled by dataclass.